### DISCUSSION TITLE
Multisite and URIs

### DISCUSSION THREAD

**Discussion by mefenlon** [statamic-rad-pack/runway#549] [Help]
### Description

I am attempting to run a multisite, each site using a model Title with differing scopes. But the routes will be different on each site frontend.

```
// config/runway.php
            'route' => '/jazz-collection/titles/{{ slug }}', //site_one
            'route' => '/titles/{{ slug }}', //site_two

```

Site one: scoped to single project
Site two: scoped to any project

```
    public function scopeRunwayListing($query)
    {
        if(Site::selected()->handle() == 'site_one')
            return $query->whereRelation('projects', 'name', 'Jazz');
        if(Site::selected()->handle() == 'site_two')
            return $query->has('projects');
    }
```
What is the best way of generating and retreiving uris for each site?

Would you accept a PR that included a scope for generating uris?
Something along the lines of
`        $query->when($query->hasNamedScope('runwayMultisite'), fn ($query) => $query->runwayMultisite());
`
and adding a column to the _runway_uris_ table like how you suggest in the [multisite docs](https://runway.duncanmcclean.com/kb-articles/multisite)?

Looking at runway/src/Routing/Traits/RunwayRoutes.php
```
   public function runwayUri(): MorphOne
    {
        //scope here
        return $this->morphOne(RunwayUri::class, 'model');
    }

    public static function bootRunwayRoutes()
    {
        static::saved(function ($model) {
            $resource = Runway::findResourceByModel($model);

            if (! $resource->hasRouting()) {
                return;
            }
            //per site logic here
            $uri = Antlers::parser()
                ->parse($resource->route(), $model->toAugmentedArray())
                ->__toString();

            $uri = Str::start($uri, '/');

            if ($model->runwayUri()->exists()) {
                //scope here
                $model->runwayUri()->first()->update(['uri' => $uri]);
            } else {
                $model->runwayUri()->create(['uri' => $uri]);
            }

            app(Cacher::class)->invalidateUrl($uri);

            app(Cacher::class)->invalidateUrls(
                Arr::get(config('statamic.static_caching.invalidation.rules'), "runway.{$resource->handle()}.urls")
            );
        });

        static::deleting(function ($model) {
            if ($model->runwayUri) {
                $model->runwayUri()->delete();
            }
        });
    }
```

### Environment

Environment
Application Name: Statamic
Laravel Version: 11.14.0
PHP Version: 8.2.19
Composer Version: 2.7.6
Environment: local
Debug Mode: ENABLED
URL: local.test
Maintenance Mode: OFF
Timezone: UTC
Locale: en

Cache
Config: NOT CACHED
Events: NOT CACHED
Routes: NOT CACHED
Views: CACHED

Drivers
Broadcasting: log
Cache: redis
Database: mysql
Logs: stack / single
Mail: smtp
Queue: database
Session: file

Filament
Packages: filament, forms, notifications, support, tables
Version: v3.2.92
Views: NOT PUBLISHED

Livewire
Livewire: v3.5.1

Spatie Permissions
Features Enabled: Default
Version: 6.9.0

Statamic
Addons: 2
Sites: 3 (site_one, site_two, site_three)
Stache Watcher: Disabled
Static Caching: Disabled
Version: 5.12.0 PRO

Statamic Addons
jacksleight/statamic-bard-texstyle: 3.2.2
statamic-rad-pack/runway: 7.5.2

**Comment by duncanmcclean** [comment]
Like [the docs mention](https://runway.duncanmcclean.com/kb-articles/multisite), Runway doesn't support the concept of "localizations" in the same way that entries of taxonomy terms do.

This is since we don't want to force developers to add columns to their database tables (not everyone needs stuff attached to sites). It also makes things a lot more complicated behind the scenes.

Because of this, I don't really want to introduce the ability to have different routes per-site to Runway, since we don't have any other multi-site functionality.

For your use case, it might be better to create custom routes in `routes/web.php`. You can still take advantage of Runway's template/layout/cascade handling though, like this:

```php
Route::get('jazz-collection/titles/{slug}', function ($slug) {
    $title = \App\Models\Title::firstWhere('slug', $slug);

    return $title->toResponse(request());
});

```

**Comment by mefenlon** [comment]
Thanks for the quick response. I understand.

If I went with the custom routes solution, would statamic static cacheing still work?

I am also looking at overriding the the route function in runway/src/Routing/Traits/RunwayRoutes.php 
Do you see any potential issues with this?

```
    public function route(): ?string
    {
        if (! $this->runwayUri) {
            return null;
        }
        if(Site::selected()->handle() == 'site_one')
            return 'site_one_prefix'.$this->runwayUri->uri;
        if(Site::selected()->handle() == 'site_two')
            return 'site_two_prefix'.$this->runwayUri->uri;
        return $this->runwayUri->uri;
    }
```

**Reply by duncanmcclean** [reply]
> If I went with the custom routes solution, would statamic static cacheing still work?

Not by default, but you can add the `StaticCache` middleware to the route, then it'll work. See the [hot tip here](https://statamic.dev/routing#laravel-routes).

> I am also looking at overriding the the route function in runway/src/Routing/Traits/RunwayRoutes.php
> Do you see any potential issues with this?

It feels a bit hacky to me and might cause unintended issues down the road, but if it fixes an issue for you, then great 👍
